src/path.c

```diff
@@ -1429,9 +1434,14 @@ int git_path_diriter_fullpath(
 
 int git_path_diriter_stat(struct stat *out, git_path_diriter *diriter)
 {
+	const char *fname;
+
 	assert(out && diriter);
 
-	return git_path_lstat(diriter->path.ptr, out);
+	fname = diriter->path.ptr + diriter->parent_len;
+	if (*fname == '/') ++fname;
+
+	return fstatat(dirfd(diriter->dir), fname, out, AT_SYMLINK_NOFOLLOW);
```

Comment:
This is a pretty big optimization. I mentioned it in https://github.com/romkatv/gitstatus/blob/master/README.md#why-fast.